src/cpp/server/server_cc.cc

```diff
@@ -1201,6 +1119,30 @@ int Server::AddListeningPort(const grpc::string& addr,
   return port;
 }
 
+void Server::Ref() {
+  shutdown_refs_outstanding_.fetch_add(1, std::memory_order_relaxed);
+}
+
+void Server::UnrefWithPossibleNotify() {
+  if (shutdown_refs_outstanding_.fetch_add(-1, std::memory_order_acq_rel) ==
+      1) {
+    // No refs outstanding means that shutdown has been initiated and no more
+    // callback requests are outstanding.
+    grpc::internal::MutexLock lock(&mu_);
+    GPR_ASSERT(shutdown_);
+    shutdown_done_ = true;
+    shutdown_done_cv_.Signal();
+  }
+}
+
+void Server::UnrefAndWaitLocked() {
+  if (shutdown_refs_outstanding_.fetch_add(-1, std::memory_order_acq_rel) ==
+      1) {
+    shutdown_done_ = true;
+  }
+  shutdown_done_cv_.WaitUntil(&mu_, [this] { return shutdown_done_; });
```

Comment:
I think line `1141` (i.e., `shutdown_done_ = true`) should be guarded by locking `mu_`. Otherwise, there is no synchronization between the thread setting `shutdown = true` and the threads reading it in the `CondVar`.

Reply:
UnrefAndWaitLocked is always called under mu_ . I've put in a comment annotation to that effect in server_impl.h and we can change that to an enforced annotation once we support that in OSS gRPC.

Reply:
Ah ok, thanks!